src/ir_pattern.py

```diff
@@ -0,0 +1,446 @@
+
+        # Split the pattern asm into 3 disjoint sets of instructions:
+        # input_refs ("in.fictive"s), body_refs, and tail_ref (the last instruction)
+        n_inputs = len(pattern.replacement_instr.inputs)
+        head_refs, tail_ref = pattern_refs[:-1], pattern_refs[-1]
+        input_refs, body_refs = head_refs[:n_inputs], head_refs[n_inputs:]
+        assert all(r.instruction.mnemonic == "in.fictive" for r in input_refs)
+        assert all(r.instruction.mnemonic != "in.fictive" for r in body_refs)
+
+        # For now, pattern inputs must be Registers, not StackLocations. It's not always
+        # trivial to create temporary StackLocations in the same way we create temporary
+        # Registers during replacement.
+        assert all(
+            isinstance(inp, Register) for inp in pattern.replacement_instr.inputs
+        )
+
+        # For now, patterns can only have 1 output put register (which must be set
+        # by the final instruction in the pattern). This simplifies pattern matching
+        # by guaranteeing that there is a there is a place where all of the (fictive)
+        # pattern inputs have been assigned and the outputs have not yet been used.
+        assert len(pattern.replacement_instr.outputs) == 1
+        assert pattern.replacement_instr.outputs == tail_ref.instruction.outputs
+
+        # Start the matching by finding all possible matches for the last instruction
+        try_matches = []
+        tail_inputs = pattern.flow_graph.instr_inputs[tail_ref]
+        for cand_ref in refs_by_mnemonic.get(tail_ref.instruction.mnemonic, []):
+            state = TryIrMatch(arch=arch)
+            cand_instr = cand_ref.instruction
+            if not state.match_ref(tail_ref, cand_ref):
+                continue
+            if not state.match_instr(tail_ref.instruction, cand_instr):
+                continue
+            states = state.permute_and_match_inputrefs(
+                tail_inputs,
+                flow_graph.instr_inputs[cand_ref],
+            )
+            try_matches.extend(states)
+
+        # Continue matching by working backwards through the pattern
+        for pat_ref in body_refs[::-1]:
+            pat_inputs = pattern.flow_graph.instr_inputs[pat_ref]
+
+            next_try_matches = []
+            for state in try_matches:
+                # By pattern construction, pat_ref should be in the state's ref_map
+                # This would be true for "disjoint" or irrelevant instructions in the
+                # pattern, like random nops.
+                cand = state.try_map_ref(pat_ref)
+                if not isinstance(cand, InstrRef):
+                    continue
+                if not state.match_instr(pat_ref.instruction, cand.instruction):
+                    continue
+                states = state.permute_and_match_inputrefs(
+                    pat_inputs,
+                    flow_graph.instr_inputs[cand],
+                )
```

Comment:
I think my plan for fixing this is to create special RefSets to represent phi registers with "the value of register X at the start of node Y". 

This makes it so that, for matching purposes, pattern inputs are matched just like internal registers.